File: chapter02/sanghyo/MovieReservationProgram/MovieReservationProgram/Model/Movie.swift
//
//  Movie.swift
//  MovieReservationProgram
//
//  Created by sanghyo on 11/29/23.
//

import Foundation

struct Movie {
    private let title: String
    private let runningTime: Int
    private let fee: Money
    private let discountPolicy: DiscountPolicy
    
    init(title: String, runningTime: Int, fee: Money, discountPolicy: DiscountPolicy) {
        self.title = title
        self.runningTime = runningTime
        self.fee = fee
        self.discountPolicy = discountPolicy
    }
    
    func getFee() -> Money {
        return self.fee
    }
    
    func calculateMovieFee(screening: Screening) -> Money {
        return fee.minus(money: discountPolicy.calculateDiscountAmount(screening: screening))
    }
}

File: chapter02/hm4725/src/Movie.cc
#include "Movie.hh"
#include <string>

/* public */
std::shared_ptr<Movie> Movie::create(
    std::string title, const std::chrono::duration<long> &runningTime,
    std::shared_ptr<Money> fee,
    std::unique_ptr<DiscountPolicy> discountPolicy) {
  struct MakeSharedEnabler : public Movie {
    MakeSharedEnabler(std::string title,
                      const std::chrono::duration<long> &runningTime,
                      std::shared_ptr<Money> fee,
                      std::unique_ptr<DiscountPolicy> discountPolicy)
        : Movie(title, runningTime, fee, std::move(discountPolicy)) {}
  };
  return std::make_shared<MakeSharedEnabler>(title, runningTime, fee,
                                             std::move(discountPolicy));
}
std::shared_ptr<Movie> Movie::getPtr() { return shared_from_this(); }

std::shared_ptr<Money> Movie::getFee() { return fee; }

std::shared_ptr<Money> Movie::calculateMovieFee(
    std::shared_ptr<Screening> screening) {
  return fee->minus(discountPolicy->calculateDiscountAmount(screening));
}

/* protected */
Movie::Movie(std::string title, const std::chrono::duration<long> &runningTime,
             std::shared_ptr<Money> fee,
             std::unique_ptr<DiscountPolicy> discountPolicy)
    : title(title),
      runningTime(runningTime),
      fee(fee),
      discountPolicy(std::move(discountPolicy)) {}